compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3066,27 +3066,37 @@ TR::Register *OMR::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::C
 
    if (SymRef && SymRef->getSymbol()->castToMethodSymbol()->isInlinedByCG())
       {
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicAdd32BitSymbol))
+      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicAddSymbol))
          {
-         return inlineAtomicMemoryUpdate(node, LADD4MemReg, cg);
+         auto op = node->getChild(1)->getDataType().isInt32() ? LADD4MemReg : LADD8MemReg;
+
+         return inlineAtomicMemoryUpdate(node, op, cg);
          }
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicAdd64BitSymbol))
+      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAddSymbol))
          {
-         return inlineAtomicMemoryUpdate(node, LADD8MemReg, cg);
+         auto op = node->getChild(1)->getDataType().isInt32() ? LXADD4MemReg : LXADD8MemReg;
+
+         return inlineAtomicMemoryUpdate(node, op, cg);
          }
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd32BitSymbol))
+      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd32BitSymbol))
          {
          return inlineAtomicMemoryUpdate(node, LXADD4MemReg, cg);
          }
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd64BitSymbol))
+      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd64BitSymbol))
          {
          return inlineAtomicMemoryUpdate(node, LXADD8MemReg, cg);
          }
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwap32BitSymbol))
+      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwapSymbol))
+         {
+         auto op = node->getChild(1)->getDataType().isInt32() ? XCHG4MemReg : XCHG8MemReg;
+
+         return inlineAtomicMemoryUpdate(node, op, cg);
+         }
+      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwap32BitSymbol))
          {
          return inlineAtomicMemoryUpdate(node, XCHG4MemReg, cg);
          }
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwap64BitSymbol))
+      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwap64BitSymbol))
```

Comment:
I will postulate that it is far better for overall code footprint that you simply collect the opcode in each `else` block and then have one single call to `inlineAtomicMemoryUpdate` after the if/else ladder.  It looks like each call takes the same arguments except for the opcode.

Reply:
Fixed in 5d7999b